src/main/java/com/google/api/gax/retrying/RetryingFutureImpl.java

```diff
@@ -0,0 +1,183 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.api.gax.retrying;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.gax.core.ApiFutureCallback;
+import com.google.common.util.concurrent.AbstractFuture;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+
+/**
+ * For internal use only.
+ *
+ * <p>
+ * This class is the key component of the retry logic. It implements the {@link RetryingFuture}
+ * facade interface, and does the following:
+ *
+ * <ul>
+ * <li>Schedules the next attempt in case of a failure using the callback chaining technique.
+ * <li>Terminates retrying process if no more retries are accepted.
+ * <li>Propagates future cancellation in both directions (from this to the attempt and from the
+ * attempt to this)
+ * </ul>
+ *
+ * This class is thread-safe.
+ */
+class RetryingFutureImpl<ResponseT> extends AbstractFuture<ResponseT>
+    implements RetryingFuture<ResponseT> {
+
+  private final Object lock = new Object();
+  private final Callable<ResponseT> callable;
+
+  private final RetryAlgorithm retryAlgorithm;
+  private final RetryingExecutor<ResponseT> retryingExecutor;
+
+  private volatile TimedAttemptSettings attemptSettings;
+  private volatile AttemptFutureCallback attemptFutureCallback;
+
+  RetryingFutureImpl(
+      Callable<ResponseT> callable,
+      RetryAlgorithm retryAlgorithm,
+      RetryingExecutor<ResponseT> retryingExecutor) {
+    this.callable = checkNotNull(callable);
+    this.retryAlgorithm = checkNotNull(retryAlgorithm);
+    this.retryingExecutor = checkNotNull(retryingExecutor);
+
+    this.attemptSettings = retryAlgorithm.createFirstAttempt();
+  }
+
+  @Override
+  public boolean cancel(boolean mayInterruptIfRunning) {
+    synchronized (lock) {
+      if (attemptFutureCallback != null) {
+        if (attemptFutureCallback.attemptFuture.cancel(mayInterruptIfRunning)) {
+          super.cancel(mayInterruptIfRunning);
+        }
+        return isCancelled();
+      } else {
+        return super.cancel(mayInterruptIfRunning);
+      }
+    }
+  }
+
+  @Override
+  public void setAttemptFuture(Future<ResponseT> attemptFuture) {
+    if (isDone()) {
+      return;
+    }
+    synchronized (lock) {
+      if (isDone()) {
+        return;
+      }
+      if (attemptFuture != null) {
+        attemptFutureCallback = new AttemptFutureCallback(attemptFuture);
+        Futures.addCallback((ListenableFuture) attemptFuture, attemptFutureCallback);
```

Comment:
*Note, all these problems existed with the previous RetriableCallable implementation (in a worse form, as activeFuture's cancelation was completelly ignored, plus the underlying call was made under the lock) , so this PR did not make things worse, just trying to mitigate the problems.

In general I don't know. There is still a problem for us, if somebody calls shutdownNow() on the underlying ScheduledExecutorService, which comes to the RetryingCallable from the outside. The problem here is that there are to many things to propagate in a reverse direction: the cancelation of java.util.concurrent executor's tasks, to guava's ListenableFutureTasks and then to RetryingFuture (plus all that custom lightweight wrappers we have). Also it requires support from the caller to actually even try to cancel the tasks returned by the shutdownNow() (most people don't do it). So multiple different layers of propagation in a reverse direction (from wrapped object to its wrapper), and as result information gets lost.

There are some heavyweight workarounds, which mitigate the issue (but still do not solve it completelly), like implementing custom shutdownNow on RetryExecutor, but that requires a lot of ugly stuff (now each task must know about RetryingFuture to cancel it when it gets canceled), and still does not solve the issue of somebody messing with aggregated pools/futures/tasks directly.